src/ssa.c

```diff
@@ -0,0 +1,1152 @@
+                int i;
+                for (i = 0; i < MAX_BB_PRED; i++) {
+                    if (!bb->prev[i].bb)
+                        continue;
+                    if (!bb->prev[i].bb->idom)
+                        continue;
+                    pred = bb->prev[i].bb;
+                    break;
+                }
+
+                for (i = 0; i < MAX_BB_PRED; i++) {
+                    if (!bb->prev[i].bb)
+                        continue;
+                    if (bb->prev[i].bb == pred)
+                        continue;
+                    if (bb->prev[i].bb->idom)
+                        pred = intersect(bb->prev[i].bb, pred);
+                }
+                if (bb->idom != pred) {
+                    bb->idom = pred;
+                    changed = 1;
+                }
+            }
+        } while (changed);
+    }
+}
+
+int dom_connect(basic_block_t *pred, basic_block_t *succ)
+{
+    if (succ->dom_prev)
+        return 0;
+
+    int i, found = 0;
+    for (i = 0; i < MAX_BB_DOM_SUCC; i++) {
+        if (pred->dom_next[i] == succ) {
+            found = 1;
+            return 0;
+        } else if (!pred->dom_next[i])
+            break;
+    }
+
+    if (i > MAX_BB_DOM_SUCC - 1) {
+        printf("Error: too many predecessors\n");
+        abort();
+    }
+
+    pred->dom_next[i++] = succ;
+    succ->dom_prev = pred;
+    return 1;
+}
+
+void bb_build_dom(fn_t *fn, basic_block_t *bb)
+{
+    basic_block_t *curr = bb;
+    while (curr != fn->bbs) {
+        if (!dom_connect(curr->idom, curr))
+            break;
+        curr = curr->idom;
+    }
+}
+
+void build_dom()
+{
+    fn_t *fn;
+    for (fn = FUNC_LIST.head; fn; fn = fn->next) {
+        fn->visited++;
+        bb_forward_traversal(fn, fn->bbs, &bb_build_dom, NULL);
+    }
+}
+
+void bb_build_df(fn_t *fn, basic_block_t *bb)
+{
+    int i, cnt = 0;
+    for (i = 0; i < MAX_BB_PRED; i++)
+        if (bb->prev[i].bb)
+            cnt++;
+
+    if (cnt > 1)
+        for (i = 0; i < MAX_BB_PRED; i++)
+            if (bb->prev[i].bb) {
+                basic_block_t *curr;
+                for (curr = bb->prev[i].bb; curr != bb->idom; curr = curr->idom)
+                    curr->DF[curr->df_idx++] = bb;
+            }
+}
+
+void build_df()
+{
+    fn_t *fn;
+    for (fn = FUNC_LIST.head; fn; fn = fn->next) {
+        fn->visited++;
+        bb_forward_traversal(fn, fn->bbs, NULL, &bb_build_df);
+    }
+}
+
+int check_killed(basic_block_t *bb, var_t *var)
+{
+    int i;
+    for (i = 0; i < bb->live_kill_idx; i++)
+        if (bb->live_kill[i] == var)
+            return 1;
+    return 0;
+}
+
+void add_killed_var(basic_block_t *bb, var_t *var)
+{
+    int i, found = 0;
+    for (i = 0; i < bb->live_kill_idx; i++)
+        if (bb->live_kill[i] == var) {
+            found = 1;
+            break;
+        }
+
+    if (found)
+        return;
+
+    bb->live_kill[bb->live_kill_idx++] = var;
+}
+
+void add_killed_block(basic_block_t *bb, var_t *var)
```

Comment:
Can you unify the naming scheme? That is, use prefix `bb_`.

Reply:
That function is for `var` to record which block kill itself. Is it better to swap the arguments?

Reply:
Up to you. I just concern the naming scheme, avoiding potential conflicts.